Android binder driver information

<!--
Before reporting a new issue, make sure that we do not have any duplicates already open.
If there is one it might be good to take part in the discussion there.

Please make sure you have checked that the issue persists on LATEST pwndbg version.

Below is a template for BUG REPORTS.
Don't include it if this is a FEATURE REQUEST.
-->


### Description

<!--
Briefly describe the problem you are having in a few paragraphs.
-->
```
pwndbg> binder
This command is a work in progress and may not work as expected.
'binder': Show Android Binder information
Traceback (most recent call last):
  File "/home/uuu/repo/pwndbg/pwndbg/commands/__init__.py", line 189, in __call__
    return self.function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/__init__.py", line 317, in _OnlyWhenQemuKernel
    return function(*a, **kw)
           ^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/__init__.py", line 396, in _OnlyWithKernelDebugSyms
    return function(*a, **kw)
           ^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/__init__.py", line 410, in _OnlyWhenPagingEnabled
    return function(*a, **kw)
           ^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 383, in binder
    bv.visit()
  File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 229, in visit
    print(self.format_proc(proc))
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 251, in format_proc
    res.append(self._format_fields(proc, fields))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 224, in _format_fields
    res.append(self._format_field(field, obj[field], only_heading=only_heading))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 151, in _format_field
    value, num_elts = self.format_rb_tree(field, value)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 177, in format_rb_tree
    for entry in for_each_rb_entry(
                 ^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/gdblib/kernel/rbtree.py", line 30, in for_each_rb_entry
    node = rb_next(node)
           ^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/gdblib/kernel/rbtree.py", line 81, in rb_next
    if rb_empty_node(node):
       ^^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/gdblib/kernel/rbtree.py", line 72, in rb_empty_node
    return int(node["__rb_parent_color"]) == int(node.address)
                                             ^^^^^^^^^^^^^^^^^
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'

If that is an issue, you can report it on https://github.com/pwndbg/pwndbg/issues
(Please don't forget to search if it hasn't been reported before)
To generate the report and open a browser, you may run `bugreport --run-browser`
PS: Pull requests are welcome
```

### Steps to reproduce

<!--
What do we have to do to reproduce the problem?
If this is connected to particular C/asm code,
please provide the smallest C code that reproduces the issue.
-->
1. clone https://github.com/gsingh93/linux-exploit-dev-env and get ubuntu rootfs from https://github.com/gsingh93/linux-exploit-dev-env/releases/tag/2024.04.07-b21e193 and clone  https://github.com/brookate/binderubuntu
2. into linux-exploit-dev-env , type `ACK=1 make linux` to build Android Common Kernel, 
3. then startup ubuntu rootfs with gdb `ACK=1 GDB=1 ROOTFS=ubuntuxx.img make run`
4. open new terminal, into linux-exploit-dev-env, start `scripts/gdb.sh` then `c` 
5.  scp binderubuntu to qemu then install gcc , g++, make , compile it (modify Makefile to skip compile driver binder, asheme)
6. run servicemanager 
7. run test_server
8. at gdb terminal, type binder to show binder information to trigger. 

Hi, @gsingh93 can you fix it if possible?

### My setup

<!--
Show us your gdb/python/pwndbg/OS/IDA Pro version (depending on your case).

NOTE: We are currently supporting only Ubuntu installations.
It is known that pwndbg is not fully working e.g. on Arch Linux (the heap stuff is not working there).
If you would like to change this situation - help us improving pwndbg and supporting other distros!

This can be displayed in pwndbg through `version` command.

If it is somehow unavailable, use:
* `show version` - for gdb
* `py import sys; print(sys.version)` - for python
* pwndbg version/git commit id
-->

can't work

```
pwndbg> binder
This command is a work in progress and may not work as expected.
Traceback (most recent call last):
  File "/home/uuu/repo/pwndbg/pwndbg/commands/__init__.py", line 183, in __call__
    return self.function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/__init__.py", line 294, in _OnlyWhenQemuKernel
    return function(*a, **kw)
           ^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/__init__.py", line 373, in _OnlyWithKernelDebugSyms
    return function(*a, **kw)
           ^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/__init__.py", line 387, in _OnlyWhenPagingEnabled
    return function(*a, **kw)
           ^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 389, in binder
    bv.visit()
  File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 231, in visit
    print(self.format_proc(proc))
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 253, in format_proc
    res.append(self._format_fields(proc, fields))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 226, in _format_fields
    res.append(self._format_field(field, obj[field], only_heading=only_heading))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 117, in _format_field
    if t.code == pwndbg.dbg_mod.TypeCode.INT:
       ^^^^^^
  File "/home/uuu/repo/pwndbg/pwndbg/dbg/gdb/__init__.py", line 994, in code
    assert self.inner.code in GDBType.CODE_MAPPING, "missing mapping for type code"
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: missing mapping for type code

```

> please check
> 
> Traceback (most recent call last): File "/home/uuu/repo/pwndbg/pwndbg/commands/**init**.py", line 183, in **call** return False ^^^^^^ File "/home/uuu/repo/pwndbg/pwndbg/commands/**init**.py", line 311, in _OnlyWhenQemuKernel File "/home/uuu/repo/pwndbg/pwndbg/commands/**init**.py", line 390, in _OnlyWithKernelDebugSyms File "/home/uuu/repo/pwndbg/pwndbg/commands/**init**.py", line 404, in _OnlyWhenPagingEnabled File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 389, in binder File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 231, in visit File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 253, in format_proc return "\n".join(res) ^^^^^^^ File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 226, in _format_fields File "/home/uuu/repo/pwndbg/pwndbg/commands/binder.py", line 117, in _format_field if t.code == name_to_code["TYPE_CODE_INT"]: ^^^^^^ File "/home/uuu/repo/pwndbg/pwndbg/dbg/gdb/**init**.py", line 994, in code AssertionError: missing mapping for type code

I tested at PR #2632, but at dev branch, it works.

--- a/pwndbg/aglib/kernel/rbtree.py
+++ b/pwndbg/aglib/kernel/rbtree.py
@@ -13,6 +13,7 @@ rb_node_type: pwndbg.dbg_mod.Type = None
 
 
 # TODO: fix `NEW_MODULE` probably this event is wrong here
+# TODO: FIXY
 @pwndbg.dbg.event_handler(EventType.NEW_MODULE)
 def init():
     global rb_root_type, rb_node_type

--- a/pwndbg/aglib/kernel/rbtree.py
+++ b/pwndbg/aglib/kernel/rbtree.py
@@ -12,10 +12,7 @@ rb_root_type: pwndbg.dbg_mod.Type = None
 rb_node_type: pwndbg.dbg_mod.Type = None
 
 
-# TODO: fix `NEW_MODULE` probably this event is wrong here
-# TODO: FIXY
 @pwndbg.dbg.event_handler(EventType.START)
-@pwndbg.dbg.event_handler(EventType.NEW_MODULE)
 def init():
     global rb_root_type, rb_node_type
     rb_root_type = pwndbg.aglib.typeinfo.load("struct rb_root")

--- a/pwndbg/aglib/kernel/rbtree.py
+++ b/pwndbg/aglib/kernel/rbtree.py
@@ -14,6 +14,7 @@ rb_node_type: pwndbg.dbg_mod.Type = None
 
 # TODO: fix `NEW_MODULE` probably this event is wrong here
 # TODO: FIXY
+@pwndbg.dbg.event_handler(EventType.START)
 @pwndbg.dbg.event_handler(EventType.NEW_MODULE)
 def init():
     global rb_root_type, rb_node_type

--- a/pwndbg/commands/binder.py
+++ b/pwndbg/commands/binder.py
@@ -143,7 +143,7 @@ class BinderVisitor:
                 if typename == "spinlock":
                     value = self.format_spinlock(value).strip()
                 elif typename == "atomic_t":
-                    value = value["counter"]
+                    value = value["counter"].value_to_human_readable()
                 elif typename == "rb_root":
                     assert field is not None
                     value, num_elts = self.format_rb_tree(field, value)
@@ -223,7 +223,7 @@ class BinderVisitor:
     ) -> str:
         res = []
         for field in fields:
-            res.append(self._format_field(field, obj[field].value_to_human_readable(), only_heading=only_heading))
+            res.append(self._format_field(field, obj[field], only_heading=only_heading))
         return "\n".join(res)
 
     def visit(self):
@@ -338,7 +338,7 @@ class BinderVisitor:
 
     def format_work(self, work: pwndbg.dbg_mod.Value) -> str:
         res = []
-        res.append(self._format_heading("binder_work", str(work["type"]), int(work.address)))
+        res.append(self._format_heading("binder_work", work["type"].value_to_human_readable(), int(work.address)))
 
         t = int(work["type"])
         # TODO: Create enum

--- a/pwndbg/commands/binder.py
+++ b/pwndbg/commands/binder.py
@@ -233,7 +233,7 @@ class BinderVisitor:
 
     def format_proc(self, proc: pwndbg.dbg_mod.Value, only_heading=False):
         res = []
-        res.append(self._format_heading("binder_proc", "PID %s" % proc["pid"], int(proc)))
+        res.append(self._format_heading("binder_proc", "PID %s" % proc["pid"].value_to_human_readable(), int(proc)))
 
         if only_heading:
             return "\n".join(res)
@@ -256,7 +256,7 @@ class BinderVisitor:
 
     def format_thread(self, thread: pwndbg.dbg_mod.Value, only_heading: bool = False) -> str:
         res = []
-        res.append(self._format_heading("binder_thread", "PID %s" % thread["pid"], int(thread)))
+        res.append(self._format_heading("binder_thread", "PID %s" % thread["pid"].value_to_human_readable(), int(thread)))
 
         if only_heading:
             return "\n".join(res)
@@ -277,7 +277,7 @@ class BinderVisitor:
         res = []
         res.append(
             self._format_heading(
-                "binder_transaction", "ID %s" % str(transaction["debug_id"]), int(transaction)
+                "binder_transaction", "ID %s" % transaction["debug_id"].value_to_human_readable(), int(transaction)
             )
         )
 
@@ -325,7 +325,7 @@ class BinderVisitor:
 
     def format_ref(self, ref: pwndbg.dbg_mod.Value, only_heading: bool = False) -> str:
         res = []
-        res.append(self._format_heading("binder_ref", "HANDLE %s" % ref["data"]["desc"], int(ref)))
+        res.append(self._format_heading("binder_ref", "HANDLE %s" % ref["data"]["desc"].value_to_human_readable(), int(ref)))
 
         if only_heading:
             return "\n".join(res)

--- a/pwndbg/commands/binder.py
+++ b/pwndbg/commands/binder.py
@@ -116,6 +116,8 @@ class BinderVisitor:
 
             if t.code == pwndbg.dbg_mod.TypeCode.INT:
                 value = int(value)
+            elif t.code == pwndbg.dbg_mod.TypeCode.BOOL:
+                value = True if int(value) else False
             elif t.code == pwndbg.dbg_mod.TypeCode.POINTER:
                 typename = t.target().name_identifier
                 if int(value) == 0:
@@ -233,7 +235,11 @@ class BinderVisitor:
 
     def format_proc(self, proc: pwndbg.dbg_mod.Value, only_heading=False):
         res = []
-        res.append(self._format_heading("binder_proc", "PID %s" % proc["pid"].value_to_human_readable(), int(proc)))
+        res.append(
+            self._format_heading(
+                "binder_proc", "PID %s" % proc["pid"].value_to_human_readable(), int(proc)
+            )
+        )
 
         if only_heading:
             return "\n".join(res)
@@ -256,7 +262,11 @@ class BinderVisitor:
 
     def format_thread(self, thread: pwndbg.dbg_mod.Value, only_heading: bool = False) -> str:
         res = []
-        res.append(self._format_heading("binder_thread", "PID %s" % thread["pid"].value_to_human_readable(), int(thread)))
+        res.append(
+            self._format_heading(
+                "binder_thread", "PID %s" % thread["pid"].value_to_human_readable(), int(thread)
+            )
+        )
 
         if only_heading:
             return "\n".join(res)
@@ -277,7 +287,9 @@ class BinderVisitor:
         res = []
         res.append(
             self._format_heading(
-                "binder_transaction", "ID %s" % transaction["debug_id"].value_to_human_readable(), int(transaction)
+                "binder_transaction",
+                "ID %s" % transaction["debug_id"].value_to_human_readable(),
+                int(transaction),
             )
         )
 
@@ -325,7 +337,11 @@ class BinderVisitor:
 
     def format_ref(self, ref: pwndbg.dbg_mod.Value, only_heading: bool = False) -> str:
         res = []
-        res.append(self._format_heading("binder_ref", "HANDLE %s" % ref["data"]["desc"].value_to_human_readable(), int(ref)))
+        res.append(
+            self._format_heading(
+                "binder_ref", "HANDLE %s" % ref["data"]["desc"].value_to_human_readable(), int(ref)
+            )
+        )
 
         if only_heading:
             return "\n".join(res)
@@ -338,7 +354,11 @@ class BinderVisitor:
 
     def format_work(self, work: pwndbg.dbg_mod.Value) -> str:
         res = []
-        res.append(self._format_heading("binder_work", work["type"].value_to_human_readable(), int(work.address)))
+        res.append(
+            self._format_heading(
+                "binder_work", work["type"].value_to_human_readable(), int(work.address)
+            )
+        )
 
         t = int(work["type"])
         # TODO: Create enum

--- a/pwndbg/commands/binder.py
+++ b/pwndbg/commands/binder.py
@@ -223,7 +223,7 @@ class BinderVisitor:
     ) -> str:
         res = []
         for field in fields:
-            res.append(self._format_field(field, obj[field], only_heading=only_heading))
+            res.append(self._format_field(field, obj[field].value_to_human_readable(), only_heading=only_heading))
         return "\n".join(res)
 
     def visit(self):

--- a/pwndbg/dbg/__init__.py
+++ b/pwndbg/dbg/__init__.py
@@ -607,6 +607,7 @@ class TypeCode(Enum):
     INT = 6
     ENUM = 7
     FUNC = 8
+    BOOL = 9
 
 
 class TypeField:

--- a/pwndbg/dbg/gdb/__init__.py
+++ b/pwndbg/dbg/gdb/__init__.py
@@ -1067,7 +1067,10 @@ class GDBValue(pwndbg.dbg_mod.Value):
     @property
     @override
     def address(self) -> pwndbg.dbg_mod.Value | None:
-        return GDBValue(self.inner.address)
+        val = self.inner.address
+        if val is None:
+            return val
+        return GDBValue(val)
 
     @property
     @override

--- a/pwndbg/dbg/gdb/__init__.py
+++ b/pwndbg/dbg/gdb/__init__.py
@@ -941,6 +941,7 @@ class GDBCommandHandle(pwndbg.dbg_mod.CommandHandle):
 
 class GDBType(pwndbg.dbg_mod.Type):
     CODE_MAPPING = {
+        gdb.TYPE_CODE_BOOL: pwndbg.dbg_mod.TypeCode.INT,
         gdb.TYPE_CODE_INT: pwndbg.dbg_mod.TypeCode.INT,
         gdb.TYPE_CODE_UNION: pwndbg.dbg_mod.TypeCode.UNION,
         gdb.TYPE_CODE_STRUCT: pwndbg.dbg_mod.TypeCode.STRUCT,

--- a/pwndbg/dbg/gdb/__init__.py
+++ b/pwndbg/dbg/gdb/__init__.py
@@ -995,6 +995,7 @@ class GDBType(pwndbg.dbg_mod.Type):
             assert self.inner.code in GDBType.CODE_MAPPING, "missing mapping for type code"
             return GDBType.CODE_MAPPING[self.inner.code]
         except Exception:
+            print(f'Invalid code {self.inner.code}-{str(self.inner)}')
             # TODO: log invalid types
             return pwndbg.dbg_mod.TypeCode.INVALID
 

--- a/pwndbg/dbg/gdb/__init__.py
+++ b/pwndbg/dbg/gdb/__init__.py
@@ -941,7 +941,7 @@ class GDBCommandHandle(pwndbg.dbg_mod.CommandHandle):
 
 class GDBType(pwndbg.dbg_mod.Type):
     CODE_MAPPING = {
-        gdb.TYPE_CODE_BOOL: pwndbg.dbg_mod.TypeCode.INT,
+        gdb.TYPE_CODE_BOOL: pwndbg.dbg_mod.TypeCode.BOOL,
         gdb.TYPE_CODE_INT: pwndbg.dbg_mod.TypeCode.INT,
         gdb.TYPE_CODE_UNION: pwndbg.dbg_mod.TypeCode.UNION,
         gdb.TYPE_CODE_STRUCT: pwndbg.dbg_mod.TypeCode.STRUCT,
@@ -996,7 +996,7 @@ class GDBType(pwndbg.dbg_mod.Type):
             assert self.inner.code in GDBType.CODE_MAPPING, "missing mapping for type code"
             return GDBType.CODE_MAPPING[self.inner.code]
         except Exception:
-            print(f'Invalid code {self.inner.code}-{str(self.inner)}')
+            print(f"Invalid code {self.inner.code}-{str(self.inner)}")
             # TODO: log invalid types
             return pwndbg.dbg_mod.TypeCode.INVALID
 

--- a/pwndbg/dbg/gdb/__init__.py
+++ b/pwndbg/dbg/gdb/__init__.py
@@ -1071,7 +1071,7 @@ class GDBValue(pwndbg.dbg_mod.Value):
     def address(self) -> pwndbg.dbg_mod.Value | None:
         val = self.inner.address
         if val is None:
-            return val
+            return None
         return GDBValue(val)
 
     @property

--- a/pwndbg/dbg/gdb/__init__.py
+++ b/pwndbg/dbg/gdb/__init__.py
@@ -996,7 +996,6 @@ class GDBType(pwndbg.dbg_mod.Type):
             assert self.inner.code in GDBType.CODE_MAPPING, "missing mapping for type code"
             return GDBType.CODE_MAPPING[self.inner.code]
         except Exception:
-            print(f"Invalid code {self.inner.code}-{str(self.inner)}")
             # TODO: log invalid types
             return pwndbg.dbg_mod.TypeCode.INVALID
 

--- a/pwndbg/dbg/lldb/__init__.py
+++ b/pwndbg/dbg/lldb/__init__.py
@@ -307,8 +307,11 @@ def map_type_code(type: lldb.SBType) -> pwndbg.dbg_mod.TypeCode:
     if c == lldb.eTypeClassFunction:
         return pwndbg.dbg_mod.TypeCode.FUNC
 
-    f = type.GetTypeFlags()
+    basic_type = type.GetCanonicalType().GetBasicType()
+    if basic_type == lldb.eBasicTypeBool:
+        return pwndbg.dbg_mod.TypeCode.BOOL
 
+    f = type.GetTypeFlags()
     if f & lldb.eTypeIsInteger != 0:
         return pwndbg.dbg_mod.TypeCode.INT
 

